src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -86,10 +84,63 @@ private static bool ShouldIncludeInTargetTypedCompletionList(
             return isConvertible;
         }
 
-        typeConvertibilityCache[type] = CompletionUtilities.IsTypeImplicitlyConvertible(semanticModel.Compilation, type, inferredTypes);
+        typeConvertibilityCache[type] = CompletionUtilities.IsTypeImplicitlyConvertible(syntaxContext.SemanticModel.Compilation, type, syntaxContext.InferredTypes);
         return typeConvertibilityCache[type];
     }
 
+    private static bool ShouldIncludeInTargetTypedCompletionListInNamedType(ISymbol symbol, TSyntaxContext syntaxContext, Dictionary<ITypeSymbol, bool> typeConvertibilityCache)
+    {
+        // Only create target typed completion entries in the object creation context
+        if (!syntaxContext.IsObjectCreationTypeContext)
+            return false;
+
+        if (symbol.GetSymbolType() is not ITypeSymbol type)
+            return false;
+
+        if (typeConvertibilityCache.TryGetValue(type, out var isConvertible))
+            return isConvertible;
+
+        for (var i = 0; i < syntaxContext.InferredTypes.Length; i++)
+        {
+            var inferredType = syntaxContext.InferredTypes[i];
+            if (inferredType.IsArrayType())
+            {
+                while (inferredType is IArrayTypeSymbol arrayType)
+                    inferredType = arrayType.ElementType;
+            }
+            else
+            {
+                // Abstract types should not be offered in target typed completion except in array contexts
+                if (type.IsAbstract)
+                    continue;
+            }
+
+            if (inferredType.IsInterfaceType())
+            {
+                if (type == inferredType.OriginalDefinition || type.AllInterfaces.Any(static (typeInterface, inferredType) => typeInterface.OriginalDefinition == inferredType.OriginalDefinition, inferredType))
+                {
+                    isConvertible = true;
+                    break;
+                }
+            }
+            else
+            {
+                var typeToCheck = type;
+                while (typeToCheck != null && typeToCheck.OriginalDefinition != inferredType.OriginalDefinition)
+                    typeToCheck = typeToCheck.BaseType;
+
+                if (typeToCheck != null)
+                {
+                    isConvertible = true;
+                    break;
+                }
+            }
+        }
```

Comment:
extract into helper that can return true/false, and then just assign one and done to isConvertible.